### ISSUE TITLE
aws_lambda_layer_version_permission is not idempotent when more than one permission exist for a layer version

### ISSUE THREAD

**Issue by Jorge-Rodriguez** [hashicorp/terraform-provider-aws#21917]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
Terraform v1.0.8
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v3.67.0
```

### Affected Resource(s)

* aws_lambda_layer_version_permission

### Terraform Configuration Files

```hcl
resource "aws_lambda_layer_version_permission" "authorizer" {
  for_each = toset([local.inari_accounts.staging.id, local.inari_accounts.production.id])

  layer_name     = module.authorizer_layer.lambda_layer_layer_arn
  version_number = module.authorizer_layer.lambda_layer_version
  principal      = each.value
  statement_id   = "${module.authorizer_layer.lambda_layer_version}-${each.value}"
  action         = "lambda:GetLayerVersion"
}
```

### Error Output

```
╷
│ Error: error adding Lambda Layer Version Permission (layer: arn:aws:lambda:eu-west-1:***:layer:authorizer_layer, version: 52): ResourceConflictException: The statement id (52-728314938062) provided already exists. Please provide a new statement id, or remove the existing statement.
│ {
│   RespMetadata: {
│     StatusCode: 409,
│     RequestID: "988722c4-d407-4244-bd1f-e4b7b6a4207b"
│   },
│   Message_: "The statement id (52-728314938062) provided already exists. Please provide a new statement id, or remove the existing statement.",
│   Type: "User"
│ }
│ 
│   with aws_lambda_layer_version_permission.authorizer["728314938062"],
│   on apigw.tf line 55, in resource "aws_lambda_layer_version_permission" "authorizer":
│   55: resource "aws_lambda_layer_version_permission" "authorizer" {
│ 
╵
```

### Expected Behavior

No changes to the infrastructure, infrastructure is up to date

### Actual Behavior

The resources are tentatively replaced

```
  # aws_lambda_layer_version_permission.okta_authorizer["008049879029"] must be replaced
-/+ resource "aws_lambda_layer_version_permission" "okta_authorizer" {
      ~ id             = "arn:aws:lambda:eu-west-1:***:layer:okta_authorizer_layer,3" -> (known after apply)
      ~ layer_name     = "arn:aws:lambda:eu-west-1:***:layer:okta_authorizer_layer" -> (known after apply) # forces replacement
      ~ policy         = jsonencode(
            {
              - Id        = "default"
              - Statement = [
                  - {
                      - Action    = "lambda:GetLayerVersion"
                      - Effect    = "Allow"
                      - Principal = {
                          - AWS = "arn:aws:iam::008049879029:root"
                        }
                      - Resource  = "arn:aws:lambda:eu-west-1:***:layer:okta_authorizer_layer:3"
                      - Sid       = "3-008049879029"
                    },
                  - {
                      - Action    = "lambda:GetLayerVersion"
                      - Effect    = "Allow"
                      - Principal = {
                          - AWS = "arn:aws:iam::728314938062:root"
                        }
                      - Resource  = "arn:aws:lambda:eu-west-1:***:layer:okta_authorizer_layer:3"
                      - Sid       = "3-728314938062"
                    },
                ]
              - Version   = "2012-10-17"
            }
        ) -> (known after apply)
      ~ revision_id    = "e86df11f-2890-4a0b-9698-7ecc5c131c24" -> (known after apply)
      ~ statement_id   = "3-008049879029" -> (known after apply) # forces replacement
      ~ version_number = 3 -> (known after apply) # forces replacement
        # (2 unchanged attributes hidden)
    }

  # aws_lambda_layer_version_permission.okta_authorizer["728314938062"] must be replaced
-/+ resource "aws_lambda_layer_version_permission" "okta_authorizer" {
      ~ id             = "arn:aws:lambda:eu-west-1:***:layer:okta_authorizer_layer,3" -> (known after apply)
      ~ layer_name     = "arn:aws:lambda:eu-west-1:***:layer:okta_authorizer_layer" -> (known after apply) # forces replacement
      ~ policy         = jsonencode(
            {
              - Id        = "default"
              - Statement = [
                  - {
                      - Action    = "lambda:GetLayerVersion"
                      - Effect    = "Allow"
                      - Principal = {
                          - AWS = "arn:aws:iam::008049879029:root"
                        }
                      - Resource  = "arn:aws:lambda:eu-west-1:***:layer:okta_authorizer_layer:3"
                      - Sid       = "3-008049879029"
                    },
                  - {
                      - Action    = "lambda:GetLayerVersion"
                      - Effect    = "Allow"
                      - Principal = {
                          - AWS = "arn:aws:iam::728314938062:root"
                        }
                      - Resource  = "arn:aws:lambda:eu-west-1:***:layer:okta_authorizer_layer:3"
                      - Sid       = "3-728314938062"
                    },
                ]
              - Version   = "2012-10-17"
            }
        ) -> (known after apply)
      ~ principal      = "008049879029" -> "728314938062" # forces replacement
      ~ revision_id    = "e86df11f-2890-4a0b-9698-7ecc5c131c24" -> (known after apply)
      ~ statement_id   = "3-008049879029" -> (known after apply) # forces replacement
      ~ version_number = 3 -> (known after apply) # forces replacement
        # (1 unchanged attribute hidden)
    }
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

**Note** the change in principal on the last planned resource change. It would appear that terraform is comparing the second resource configuration with the first of the permissions applied to the layer version, not the second one.

**Comment by Jorge-Rodriguez** [comment]
Ping @anGie44 @Avolynsk

**Comment by fishfacemcgee** [comment]
Not sure what methods, if any, there are to "bump" an issue, but just want to report that this resource is still bugged. Going by the error, it appears the statement does not get deleted. My guess is it's related to whatever's leading the import behavior to be tied to the layer ARN and version, which is not specific enough given how granular the resource is.

**Comment by vedimuthan** [comment]
I am facing the same problem and surprised that this issue is not being addressed for more than 3 years.

**Comment by evanash1222** [comment]
Hello team, the issue can still be reproduced with 5.98.0 provider. Seems the information recorded under terraform.tfstate is not consistent with resource definition.

**Comment by holomekc** [comment]
We also have this issue. I think the issue is the read func. It has totally wrong assumptions:
https://github.com/hashicorp/terraform-provider-aws/blob/1a5ea8237f818882817fd262180c59e51a5bd3b8/internal/service/lambda/layer_version_permission.go#L130-L209

It checks for statements length > 0 and then just picks index 0. So it only ever compares to the first entry of the statements. 

I am not sure regarding what is written in the terraform state. This also looks totally messed up. The first resource contains a policy, which just contains what was added by the resource. The next resource contains a policy, which contains the content from the first resource and of the second resource, but as said, after reading the data it uses index 0 and then is confused that the values are not properly set. 

So in short everything larger than one entry for aws_lambda_layer_version_permission and a specify layer name and version is broken.

The main problem is that AWS manages one policy and the terraform setup allows to define separated resources only, which are not merged and extracted properly after apply. I guess the easiest solution would be to be closer to the actual AWS definition and allow to specify the policy directly. Then we could collect all entries for the policy first and then apply it once to one specific layer name and version.

**Comment by holomekc** [comment]
This should work as a workaround. Super ugly, but yeah. Assuming you have:

a list org_version_pairs with:
```tf
{
  id: "<unique id>"
  org: "<org-id>"
  version: "<lambda-layer-version>"
}
```

a list account_version_pairs with:
```tf
{
  id: "<unique id>"
  account: "accountId"
  version: "<lambda-layer-version>"
}
```


```tf
resource "terraform_data" "lambda_layer_permission_orgs" {
  for_each = { for pair in local.org_version_pairs : "${pair.org}-${pair.version}" => pair }

  triggers_replace = {
    statement_id = "org-access-${each.value.id}"
    org          = each.value.org
    version      = each.value.version
    layer        = var.layer_name
  }

  provisioner "local-exec" {
    command = <<EOT
      aws lambda add-layer-version-permission \
        --layer-name ${self.triggers_replace.layer} \
        --version-number ${self.triggers_replace.version} \
        --statement-id ${self.triggers_replace.statement_id} \
        --principal "*" \
        --organization-id ${self.triggers_replace.org} \
        --action lambda:GetLayerVersion
    EOT
  }

  provisioner "local-exec" {
    when    = destroy
    command = <<EOT
      aws lambda remove-layer-version-permission \
        --layer-name ${self.triggers_replace.layer} \
        --version-number ${self.triggers_replace.version} \
        --statement-id ${self.triggers_replace.statement_id} || \
      if [[ $? -eq 254 ]]; then
        echo "Permission already missing, ignoring error."
      else
        exit $?
      fi
    EOT
  }
}

resource "terraform_data" "lambda_layer_permission_accounts" {
  for_each = { for pair in local.account_version_pairs : "${pair.account}-${pair.version}" => pair }

  triggers_replace = {
    statement_id = "account-access-${each.value.id}"
    account      = each.value.account
    version      = each.value.version
    layer        = var.layer_name
  }

  provisioner "local-exec" {
    command = <<EOT
      aws lambda add-layer-version-permission \
        --layer-name ${self.triggers_replace.layer} \
        --version-number ${self.triggers_replace.version} \
        --statement-id ${self.triggers_replace.statement_id} \
        --principal ${self.triggers_replace.account} \
        --action lambda:GetLayerVersion
    EOT
  }

  provisioner "local-exec" {
    when    = destroy
    command = <<EOT
      aws lambda remove-layer-version-permission \
        --layer-name ${self.triggers_replace.layer} \
        --version-number ${self.triggers_replace.version} \
        --statement-id ${self.triggers_replace.statement_id} || \
      if [[ $? -eq 254 ]]; then
        echo "Permission already missing, ignoring error."
      else
        exit $?
      fi
    EOT
  }
}
```

But this is just an example you can adjust ids and maybe the list approach is also not suitable for your use-case. I hope this helps anyway.

**Comment by andreistavarache** [comment]
Why has this issue remained unaddressed for the past four years? While there are workarounds available, it continues to be a source of inconvenience.

**Comment by Lerdson** [comment]
@andreistavarache Not sure what the community is to do at this point. I submitted a fix October last year that still hasn't been looked at. Review should have been super low effort. Seems like one of the biggest things holding terraform back is the owners.
